llvm/lib/MC/MCCASObjectV1.cpp

```diff
@@ -421,27 +466,132 @@ Expected<SectionRef> SectionRef::create(MCCASBuilder &MB,
 }
 
+Expected<SmallVector<char, 0>> MCCASReader::applyRelocationAddends(
+    const SmallVector<cas::ObjectRef> &Refs, StringRef RelocData,
+    StringLiteral KindString, StringRef SectionName,
+    bool &IsPaddingRefPresent) {
+  if (auto E = decodeRelocations(*this, RelocData))
+    return std::move(E);
+  SmallVector<char, 0> SectionData;
+  auto RelocRef = MCObjectProxy::get(Schema, Schema.CAS.getProxy(Refs[0]));
+  if (!RelocRef)
+    return RelocRef.takeError();
+  if (auto R = RelocAddendRef::Cast(*RelocRef)) {
+    auto RelocationData = R->getData();
+    for (uint32_t I = 1; I < Refs.size(); I++) {
+      auto BlockRef = MCObjectProxy::get(Schema, Schema.CAS.getProxy(Refs[I]));
+      if (!BlockRef)
+        return BlockRef.takeError();
+      if (BlockRef->getKindString() == KindString) {
+        llvm::append_range(SectionData, BlockRef->getData());
+      } else if (auto P = PaddingRef::Cast(*BlockRef)) {
+        if (I != Refs.size() - 1)
+          return createStringError(
+              inconvertibleErrorCode(),
+              "PaddingRef must be the last reference of the Section");
+        IsPaddingRefPresent = true;
+      } else {
+        StringLiteral ErrorMsg =
+            "Every other reference to a {0}, must be a {1}, or {2} object!";
+        llvm::format(ErrorMsg.begin(), SectionName.data(), KindString.data(),
+                     PaddingRef::KindString.data());
+        return createStringError(inconvertibleErrorCode(), ErrorMsg.data());
+      }
+    }
+    uint32_t RelocDataIndex = 0;
+    for (auto Relocs : Relocations.back()) {
+      auto Size = getRelocationSize(Relocs, this->getEndian(), this->getArch());
+      if (!Size)
+        return Size.takeError();
+      if (*Size + RelocDataIndex > RelocationData.size()) {
+        return createStringError(inconvertibleErrorCode(),
+                                 "Relocation Data not encoded correctly, "
+                                 "reading past the end of the relocation data");
+      }
+      for (uint32_t I = 0; I < *Size; I++)
+        SectionData[Relocs.r_word0 + I] = RelocationData[RelocDataIndex + I];
+      RelocDataIndex += *Size;
+    }
+    return SectionData;
+  }
+  return createStringError(
+      inconvertibleErrorCode(),
+      "First reference to debug line section must be a RelocAddendRef object!");
+}
+
+Expected<uint64_t> MCCASReader::materializeDebugLineSection(
+    const SmallVector<cas::ObjectRef> &Refs, StringRef RelocData) {
+  bool IsPaddingRefPresent = false;
+  auto DebugLineData =
+      applyRelocationAddends(Refs, RelocData, DebugLineRef::KindString,
+                             "debug_line", IsPaddingRefPresent);
+  if (!DebugLineData)
+    return DebugLineData.takeError();
+  OS << *DebugLineData;
+  if (IsPaddingRefPresent) {
+    auto PadRef =
+        MCObjectProxy::get(Schema, Schema.CAS.getProxy(Refs[Refs.size() - 1]));
+    if (!PadRef)
+      return PadRef.takeError();
+    if (auto P = PaddingRef::Cast(*PadRef)) {
```

Comment:
Aren't we in an invalid state if `IsPaddingRefPresent` is true but the cast fails? If so, I'd remove the `if` and use an `assert` or return an error